vcf.c

```diff
@@ -911,29 +982,47 @@ int bcf_hdr_add_hrec(bcf_hdr_t *hdr, bcf_hrec_t *hrec)
     return hrec->type==BCF_HL_GEN ? 0 : 1;
 }
 
-/*
- *  Note that while querying of FLT,INFO,FMT,CTG lines is fast (the keys are hashed),
- *  the STR,GEN lines are searched for linearly in a linked list of all header lines.
- *  This may become a problem for VCFs with huge headers, we might need to build a
- *  dictionary for these lines as well.
- */
 bcf_hrec_t *bcf_hdr_get_hrec(const bcf_hdr_t *hdr, int type, const char *key, const char *value, const char *str_class)
 {
     int i;
     if ( type==BCF_HL_GEN )
     {
+        // e.g. ##fileformat=VCFv4.2
+        //      ##source=GenomicsDBImport
+        //      ##bcftools_viewVersion=1.16-80-gdfdb0923+htslib-1.16-34-g215d364
+        if ( value )
+        {
+            kstring_t str = {0,0,0};
+            ksprintf(&str, "##%s=%s", key,value);
+            bcf_hdr_aux_t *aux = get_hdr_aux(hdr);
+            khint_t k = kh_get(hdict, aux->gen, str.s);
+            free(str.s);
+            if ( k == kh_end(aux->gen) ) return NULL;
+            return kh_val(aux->gen, k);
+        }
```

Comment:
Obviously one benefit of storing all the unstructured keys in a hash table is fast searching.  However I'm struggling to see the use case here given, for uniqueness purposes, we hash on both key and value.

If I already know "source" is "GenomicsDBImport" then I can query it and get back that source is GenomicsDBImport.  Not particularly useful!  About all it adds is a way to probe a list of known values and see what comes back present.  So we could for example test if fileformat is 4.1, 4.2, 4.3 or 4.4, but we couldn't ask the basic question of "what is the file format version?".

Are there plans to extend the hash to have both keys and key=value components, so we could pass in NULL and return the first matching key?

Bonus points if the keys can form an internal linked list, but I'm not sure how much scope we have for doing that without breaking the ABI.  `bcf_hrec_t` has nkeys and pointers to arrays of key/val for the structured data, so there's room to store the result.    To make this transparent though it'd either need some abuse/hack of those array elements to add the internal linked list of hrecs, or use an internal struct for the values in hdict instead of hrec directly (the cleaner option).

Note: This isn't a reason to not merge this PR though.  Very much something to consider for the "to be continued" part.